{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "708b0f64",
   "metadata": {},
   "source": [
    "# Recap\n",
    "\n",
    "Data comes in different formats: `.fa.gz` for genomes, `.fastq.gz` for raw reads, `.bam` for mapped reads, `.vcf.gz` for genotype calls.\n",
    "\n",
    "* Format means that specific elements are expected in a specific order.\n",
    "* Reference genomes have different versions - always know which one you use!\n",
    "\n",
    "Command line tools to filter, specialised tools for such data: `samtools` for mapped reads, `gatk MarkDuplicates` for filtering, `gatk HaplotypeCaller` for genotype calling.\n",
    "\n",
    "Variant calls can be for all basepairs, for only SNPs, or in a condensed genome-wide format (\"gvcf\") - but all based on the convention of the `VCF` file format. `VCF` files can be for many individuals.\n",
    "\n",
    "One important toolkit to work with the files is `bcftools`.\n",
    "\n",
    "`bcftools concat` to add rows (genomic positions for the same individuals), `bcftools merge` to add columns (individuals for the same positions)\n",
    "\n",
    "`bcftools view -r ` to obtain regions\n",
    "\n",
    "`bcftools view -S ` to obtain individuals\n",
    "\n",
    "`bcftools view -a ` to trim alleles\n",
    "\n",
    "`bcftools filter` to perform a lot of filtering\n",
    "\n",
    "Now for something slightly different. We have looked at sequencing data, and often we obtain numbers. It is always a good idea to visualise some of this. How do you do this from the command line (assuming you are not on your laptop but a remote server)?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db7ed003-3088-48d4-a113-26e69b94a6d3",
   "metadata": {},
   "source": [
    "# Some data visualisation\n",
    "\n",
    "## Proportions of sequencing reads\n",
    "\n",
    "Assume you store information on your raw data in files. You have 20 sequencing experiments, and obtain values for\n",
    "\n",
    "* raw sequencing reads\n",
    "* reads after adapter removal\n",
    "* reads mapped\n",
    "* reads with sufficient quality (e.g. -m 35 -q 30)\n",
    "* reads after duplicate removal\n",
    "\n",
    "These proportions would tell you if something went wrong - and where it went wrong. A problem in library preparation may lead to high adapter abundance. DNA degradation may lead to small number of reads mapped, short reads or reads of low quality. PCR overamplification may lead to high numbers of duplicates, etc. It might be useful to look at this visually to check what is going on.\n",
    "\n",
    "Here is a table with the values: `~/bioinfo-course/notebooks/12_seq_viz/data/seqdata_summary.tsv`\n",
    "\n",
    "For visualization, working in R is nice. RStudio does unfortunately not run on the server using the current platform, so we may work on the R command line interface for this course.\n",
    "\n",
    "This also means plots are not showing up immediately, you need to save them.\n",
    "\n",
    "However, if you have RStudio on your computer, you may download the material and use it there.\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "```\n",
    "\n",
    "Now we load the table:\n",
    "\n",
    "```\n",
    "seqdata<-read.table(\"~/bioinfo-course/notebooks/12_seq_viz/data/seqdata_summaries.tsv\", sep=\"\\t\",header=T)\n",
    "```\n",
    "\n",
    "* A first thing is to get the proportions of reads retained, not calculating one by one, but at once.\n",
    "\n",
    "```\n",
    "seqdata[,-1]/seqdata[,2]\n",
    "```\n",
    "\n",
    "* Maybe easier to read as percentage?\n",
    "\n",
    "```\n",
    "round(seqdata[,-1]/seqdata[,2]*100,1)\n",
    "```\n",
    "\n",
    "As you may see, there are problematic samples around, where you lose a large proportion of reads towards the end of the tests. Some might be ancient DNA, some might be degraded, some might be overamplified with a PCR.\n",
    "\n",
    "Let's visualise it! An easy way with base R would be a barplot, showing the loss of data in each sample. For that, you need to turn it into a matrix, exchange columns and rows.\n",
    "\n",
    "* Let's do it!\n",
    "\n",
    "```\n",
    "bdata<-t(as.matrix(seqdata[,-1]))\n",
    "\n",
    "png(\"barplot.png\",800,400)\n",
    "barplot(bdata,beside=T,names.arg=seqdata[,1])\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "How about a legend?\n",
    "\n",
    "```\n",
    "png(\"barplot.png\",800,400)\n",
    "barplot(bdata,beside=T,names.arg=seqdata[,1],legend.text=colnames(seqdata)[-1])\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "Ok, this looks ugly... We may need to adapt the plot a little bit to give it enough space (`ylim`), and change the legend a bit (`args.legend` for making it horizontal and located in the middle):\n",
    "\n",
    "```\n",
    "png(\"barplot.png\",800,400)\n",
    "barplot(bdata,beside=T,names.arg=seqdata[,1],legend.text=colnames(seqdata)[-1],args.legend=list(x=\"top\",horiz=T),ylim=c(0,7200000))\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "As you can see, there are samples that visually stick out at different points. Assume this is a screening, where you want to know for a couple of samples how good they work for sequencing, you can choose which ones to continue with! This is very nice!\n",
    "\n",
    "* From what we know about aDNA, they tend to lose a lot of data at each step, this might be the case for two samples.\n",
    "\n",
    "* One sample has most data removed as unmapped - likely a sample with very little human DNA.\n",
    "\n",
    "* One sample has a sharp drop when removing low quality mappings - this might be due to ultra-short reads?\n",
    "\n",
    "* One sample loses most of the data when removing duplicates - this could be the case after over-amplification in a \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e4427b",
   "metadata": {},
   "source": [
    "## Coverage distribution\n",
    "\n",
    "Let's take the coverage from the chr18 data from a previous challenge! This is the `bcftools stats` output, and again we can use `awk` to subset the information which we need, the sequencing depths (`DP`).\n",
    "\n",
    "```\n",
    "awk '$1==\"DP\"' ~/bioinfo-course/notebooks/12_seq_viz/data/chr18stats.txt > ~/test/DP.txt\n",
    "```\n",
    "\n",
    "Note that you operate here in `bash` to run certain analyses, and in `R` to visualise. *These are different environments with different structures and commands!* You need to be aware of this, and may use two parallel Terminals.\n",
    "\n",
    "* Reading in the table the same way:\n",
    "\n",
    "```\n",
    "covdata<-read.table(\"~/test/DP.txt\", sep=\"\\t\",header=T)\n",
    "```\n",
    "\n",
    "* Oh no, another problem, the first is data, no header. What to do?\n",
    "\n",
    "Here, we only need to care about the third and the sixth column, i.e. the depth and the number of sites at each depth. We can do a line plot from this, right?\n",
    "\n",
    "```\n",
    "plot(x=covdata[,3],y=covdata[,6],type=\"l\")\n",
    "```\n",
    "\n",
    "Ok, not nice, because almost no positions are at depth >25.\n",
    "\n",
    "```\n",
    "covdata2<-covdata[which(as.numeric(covdata[,3])<=25),]\n",
    "plot(x=covdata2[,3],y=covdata2[,6],type=\"l\")\n",
    "```\n",
    "\n",
    "Much better! Finally, let's make the plot a bit nicer:\n",
    "\n",
    "* a thick line between dots\n",
    "* axis labels\n",
    "* a nice title\n",
    "\n",
    "Now, we can add the coverage distribution of another sample, as provided here:\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec3fbd12",
   "metadata": {},
   "source": [
    "## Heterozygosity\n",
    "\n",
    "Now, we may go back to the 1000 genomes VCF file we used earlier, and obtain heterozygous positions. \n",
    "\n",
    "```\n",
    "tbd\n",
    "```\n",
    "\n",
    "\n",
    "* Let's plot the results!\n",
    "\n",
    "```\n",
    "tbd\n",
    "```\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
