{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db7ed003-3088-48d4-a113-26e69b94a6d3",
   "metadata": {},
   "source": [
    "# Recap\n",
    "\n",
    "Data comes in different formats: `.fa.gz` for genomes, `.fastq.gz` for raw reads, `.bam` for mapped reads, `.vcf.gz` for genotype calls.\n",
    "\n",
    "* Format means that specific elements are expected in a specific order.\n",
    "* Reference genomes have different versions - always know which one you use!\n",
    "\n",
    "Command line tools to filter, specialised tools for such data: `samtools` for mapped reads, `gatk MarkDuplicates` for filtering, `gatk HaplotypeCaller` for genotype calling.\n",
    "\n",
    "Variant calls can be for all basepairs, for only SNPs, or in a condensed genome-wide format (\"gvcf\") - but all based on the convention of the `VCF` file format. `VCF` files can be for many individuals.\n",
    "\n",
    "One important toolkit to work with the files is `bcftools`.\n",
    "\n",
    "`bcftools concat` to add rows (genomic positions for the same individuals), `bcftools merge` to add columns (individuals for the same positions)\n",
    "\n",
    "`bcftools view -r ` to obtain regions\n",
    "\n",
    "`bcftools view -S ` to obtain individuals\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae4c8c80",
   "metadata": {},
   "source": [
    "## Quality filtering\n",
    "\n",
    "### Subsetting SNPs by quality\n",
    "\n",
    "When analysing data, we usually want to use variants that can meet some conditions, for example, variants with good quality.\n",
    "\n",
    "* As before, we may extract biallelic SNPs that passed the quality checks.\n",
    "\n",
    "```\n",
    "bcftools view chr18.FIN.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.FIN.biallelic.snps.vcf.gz &\n",
    "bcftools view chr18.IBS.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.IBS.biallelic.snps.vcf.gz\n",
    "```\n",
    "\n",
    "If we have a look at these files, we can see that they contain a lot of positions where all individuals within the population are `0/0` (or `0|0`). If were are interested in what is going on within the population, we may restrict to polymorphisms within.\n",
    "\n",
    "\n",
    "### Restricting data to polymorphic sites\n",
    "\n",
    "Looking at the [documentation](https://samtools.github.io/bcftools/bcftools.html#view) of `bcftools view`, we find the flag `-a` to trim unseen alternative alleles. In this case, one should first trim the alleles, and afterwards filter for biallelic SNPs.\n",
    "\n",
    "* Let's trim and filter, and count how much is left in each case!\n",
    "\n",
    "```\n",
    "bcftools view chr18.FIN.vcf.gz -a | bcftools view -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.FIN.true_biallelic.snps.vcf.gz &\n",
    "bcftools view chr18.IBS.vcf.gz -a | bcftools view -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.IBS.true_biallelic.snps.vcf.gz\n",
    "```\n",
    "\n",
    "* Given that the number of individuals in each dataset is similar, does it tell us something about the diversity within these populations?\n",
    "\n",
    "\n",
    "### Much more filtering\n",
    "\n",
    "Beyond the very basic filtering above, `bcftools filter` provides a variety of options of filtering. The easiest are defined with `-e` to *exclude* sites with a certain condition, and `-i` to *include* those with a certain condition.\n",
    "\n",
    "```\n",
    "bcftools filter\n",
    "```\n",
    "\n",
    "You need to provide [*expressions*](https://samtools.github.io/bcftools/bcftools.html#expressions), in a very similar way you provide *expressions* to `awk`, including comparison operators (`>=` etc), logical operators (`&&`), selection of samples with square brackets (`[5:6]`), and many more which cannot be comprehensively discussed here.\n",
    "\n",
    "* Let's try some!\n",
    "\n",
    "```\n",
    "bcftools filter '-i GT[0]==\"0|1\"' chr18.vcf.gz | less\n",
    "\n",
    "bcftools filter '-e GT[20]==\"0|0\"' chr18.vcf.gz | less\n",
    "\n",
    "```\n",
    "\n",
    "* To see what it is doing, look at:\n",
    "\n",
    "```\n",
    "bcftools filter '-e GT[20]==\"0|0\"' chr18.vcf.gz | grep -v \"^#\" | cut -f 30 | less\n",
    "\n",
    "bcftools filter '-i GT[1]==\"1|2\"' chr18.vcf.gz | less\n",
    "```\n",
    "\n",
    "* In the `INFO` column, there is a lot of information, such as allele frequencies (`AF`).\n",
    "\n",
    "```\n",
    "bcftools filter '-i INFO/AF>0.3' chr18.vcf.gz | cut -f 8 | less\n",
    "```\n",
    "\n",
    "The 1000 genomes file is particularly reduced, containing only genotypes. We may also try it on a more \"raw\" file from a previous session or the challenge.\n",
    "\n",
    "* Just to try out some:\n",
    "\n",
    "```\n",
    "bcftools filter '-i GT==\"./.\"' chr18_a.vcf.gz | less\n",
    "\n",
    "bcftools filter '-i FORMAT/DP>5' chr18_a.vcf.gz | less \n",
    "\n",
    "bcftools filter '-e FORMAT/DP<5 && FORMAT/DP>8' chr18_a.vcf.gz | less \n",
    "```\n",
    "\n",
    "With these types of commands, you can get a nice and clean dataset of genotypes.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d66aaffc",
   "metadata": {},
   "source": [
    "## Now for something slightly different.\n",
    "\n",
    "We have looked at sequencing data, and often we obtain numbers. It is always a good idea to visualise some of this. How do you do this from the command line (assuming you are not on your laptop but a remote server)?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be898e39",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
