{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db7ed003-3088-48d4-a113-26e69b94a6d3",
   "metadata": {},
   "source": [
    "# Global recapitulation\n",
    "\n",
    "## BASH (& R)\n",
    "\n",
    "* We use the command line (on a remote server), with a hierarchical file system: *directory/subdirectory/file.extension*.\n",
    "\n",
    "* We execute *commands* with *-arguments* on *.files* (syntax). Commands take *input* and give *output*.\n",
    "\n",
    "* Files are often compressed, marked extension such as `.gz`.\n",
    "\n",
    "* We can re-use commands (*Ctrl-C Ctrl-V!*; scrolling with cursor arrows), and autocomplete with *TAB* key.\n",
    "\n",
    "* This is not secret knowledge, but *open source*: most problems have been encountered before, so use *google* etc.\n",
    "\n",
    "* Basic commands for navigation: `cd `, `ls`, `pwd` etc.\n",
    "\n",
    "* Basic commands to look at files, copy, remove: `less`, `wc`, `head`, `cp`, `rm` etc.\n",
    "\n",
    "* Basic commands to subset/modify files: `grep`, `cut`, `awk`, `sort` etc.\n",
    "\n",
    "* Regular expressions allow complex arguments: `==`, `>=`, `&&`, `[0-9]`, `^` etc.\n",
    "\n",
    "* We use a github repository for the code.\n",
    "\n",
    "* `R` can be used for statistics and visualisation (even as command line tool)!\n",
    "\n",
    "* QC of raw sequencing stats, filtering outcomes.\n",
    "\n",
    "* For example: Distribution of coverage is good to look at!\n",
    "\n",
    "\n",
    "## Sequencing data\n",
    "\n",
    "* A genome comes as fasta `.fa` - looks like: header, sequence (possibly many rows)\n",
    "\n",
    "* Raw sequencing data comes as `.fastq` format - header, sequence, spaceholder, quality scores; *always* repeat of 4 rows per individual sequencing read (possibly billions of rows)\n",
    "\n",
    "* Sequencing data has adapters, remove them (`trimmomatic`)! Check for quality (`fastQC`)!\n",
    "\n",
    "* Mapping: finding the right spot in the reference genome `.fa` for each read in `.fastq` with `bwa`.\n",
    "\n",
    "* Mapped sequencing data in `.bam` (binary compressed) format - several header rows, then *always* one line per read  with position, sequence, quality scores etc.\n",
    "\n",
    "* Program to analyze, generate statistics, filter: `samtools view`, `samtools sort`, `samtools stats` etc.\n",
    "\n",
    "* PCR duplicates to be removed: `gatk MarkDuplicates`.\n",
    "\n",
    "* Sequencing coverage: `mosdepth`.\n",
    "\n",
    "\n",
    "## Genotype data\n",
    "\n",
    "* Genotype calling: determine the allele state for each position in the genome (`.fa`) for mapped sequencing reads (`.bam`).\n",
    "\n",
    "* Program: `gatk HaplotypeCaller`, standard for human genotype calling; `-ERC` to determine if all sites, only variant sites, or genome-wide data.\n",
    "\n",
    "* Genotypes are in `vcf` format, with multiple header rows, then *always* one line per genotype (in \"gvcf\" file - possibly region). Nine columns with info (position, alleles), from column 10 on genotypes.\n",
    "\n",
    "* Program to analyze, generate statistics, filter: `bcftools`\n",
    "\n",
    "* `bcftools view` for regions, individuals, alleles\n",
    "\n",
    "* `bcftools concat` across positions\n",
    "\n",
    "* `bcftools merge` across individuals\n",
    "\n",
    "* `bcftools filter` for many possible things\n",
    "\n",
    "* Possibly 1000s of individuals in one `vcf` file!\n",
    "\n",
    "* With that, you can do population genetics!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fdbd3c6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
