{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db7ed003-3088-48d4-a113-26e69b94a6d3",
   "metadata": {},
   "source": [
    "# Recap\n",
    "\n",
    "## Use Google! Use --help!\n",
    "\n",
    "You know how to navigate: `cd `, `less`, `zcat`, `mv`, `rm` etc.\n",
    "\n",
    "You know how to run commands on the command line, and how they handle data: `|` `>` - and also, what happens with `.gz` files!\n",
    "\n",
    "You know basic tools: `grep`, `cut`, `awk` and some special characters and regular expressions:  `*`, `==`, `&&`, `[0-9]` etc.\n",
    "\n",
    "You know what `.fa` files look like and mean, what `.fastq` files look like and mean.\n",
    "\n",
    "You can clean them up by doing adapter trimming.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fe2baab7-7988-481e-9893-a10ed5d83de0",
   "metadata": {},
   "source": [
    "# Now: Mapping of sequencing data\n",
    "\n",
    "![image.png](attachment:image.png)\n",
    "\n",
    "Finding where in the genome a short sequencing fragment belongs.\n",
    "\n",
    "Doing that for millions of lines is a challenge - algorithmic abbreviations. The most commonly used command line tool is called `bwa`, find [here](https://bio-bwa.sourceforge.net/bwa.shtml) the documentation. \n",
    "\n",
    "## **BWA**\n",
    "\n",
    "This tool has several different algorithms. The most commonly used one is `bwa mem`. Note that here we use that one because it is the easiest, but the sample data is (semi-)ancient, and in practice a different algorithm would be favourable.\n",
    "\n",
    "You provide a reference genome to align to. Actually, this one has to be *indexed* first, meaning a \"guide\" to the \"map\" is created. This \"guide\" looks the same for a given reference genome, so it just needs to be created once, and you can [download](https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/analysisSet/) a ready-made index. It is already in the same directory as the reference genome.\n",
    "\n",
    "Then, you point to the index and map your fastq file:\n",
    "\n",
    "```\n",
    "bwa mem -R '@RG\\tID:1\\tSM:1' /home/local/ANTHROPOLOGY/kuhlwilmm83/refgen/hg19/hg19.p13.plusMT.no_alt_analysis_set.fa test_p.fastq.gz > test.sam\n",
    "```\n",
    "\n",
    "Output: **SAM file** - a file that contains the sequence and its coordinates on the 23 chromosomes.\n",
    "\n",
    "**Note**: Here, we use a SAM file entirely for demonstration purpose. If you ever work on real data, *never* create a SAM file! It is extremely inefficient, and the output of bwa should immediately piped into another tool in order to get a BAM file (compressed).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e78fc882-7f45-41c0-8016-a2c848794280",
   "metadata": {},
   "source": [
    "# SAM/BAM files\n",
    "\n",
    "Now, we have created a sequencing alignment file:\n",
    "\n",
    "![image.png](attachment:image.png)\n",
    "\n",
    "It contains more than just the sequence information:\n",
    "\n",
    "* Read information\n",
    "* Position in the genome\n",
    "* Mapping quality\n",
    "* Information from FASTQ (Sequence & quality)\n",
    "* CIGAR (alignment)\n",
    "* Other stuff\n",
    "\n",
    "**BAM** or **CRAM**: compressed versions of such files, and much more efficient! However, they are different from zipped (`.gz`) files, but *binary* files instead. That means, they use a different, more basal encoding, and cannot be accessed by with usual text processing tools such as `grep`.\n",
    "\n",
    "How do we get such a file? More generally, how do we process and access SAM/BAM/CRAM files?\n",
    "\n",
    "## [**SAMTOOLS**](http://www.htslib.org/doc/samtools.html)\n",
    "\n",
    "Let's first save some space on the server. \n",
    "\n",
    "```\n",
    "samtools view -b -o test.bam test.sam\n",
    "\n",
    "ls -lh test.*\n",
    "```\n",
    "\n",
    "Now you see the difference in space! Considering the significant increase in information in the file, also compare to the raw (zipped!) sequencing data. Now remove `test.sam`!\n",
    "\n",
    "And now you see that you cannot access it as text file anymore: `less test.bam` or `zless test.bam` - nope!\n",
    "\n",
    "So, from here, you have to rely on `samtools` to see what is going on.\n",
    "\n",
    "```\n",
    "samtools view test.bam\n",
    "```\n",
    "\n",
    "**STOP!**\n",
    "\n",
    "```\n",
    "samtools view test.bam | less\n",
    "```\n",
    "\n",
    "The file should have a header, but by default is it not shown.\n",
    "\n",
    "```\n",
    "samtools view -h test.bam | less\n",
    "```\n",
    "\n",
    "Let's inspect this file a bit!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c80cb124",
   "metadata": {},
   "source": [
    "## Something to do with SAM files\n",
    "\n",
    "Now we can try to solve some tasks!\n",
    "\n",
    "* How many reads are there?\n",
    "\n",
    "* How many reads were mapped to chr21?\n",
    "\n",
    "It will be useful to sort these mapped reads so that one can find reads in the same region more easily.\n",
    "\n",
    "```\n",
    "samtools sort -o test.sorted.bam test.bam\n",
    "```\n",
    "\n",
    "In practice, one can already do that earlier, directly from the output of `bwa`, e.g. `bwa mem genome.fa data.fastq | samtools sort | samtools view -b -o data.sorted.bam`.\n",
    "\n",
    "We can then also create a \"guide\" for bam files, to make it easier for other programs to find things:\n",
    "\n",
    "```\n",
    "samtools index sample.sorted.bam\n",
    "```\n",
    "\n",
    "This is important later on. For now, `samtools` does have a couple of more functions, you can explore them based on the manual. As the last step now, we may calculate some statistics with `samtools stats`!\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec3fbd12",
   "metadata": {},
   "source": [
    "# Now you are ready for the second Challenge!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fdbd3c6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
