{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db7ed003-3088-48d4-a113-26e69b94a6d3",
   "metadata": {},
   "source": [
    "# Recap\n",
    "\n",
    "Raw sequencing data (`.fastq.gz`) mapped to a reference genome (`.fa.gz`) with `bwa`, creating files with mapped reads(`.bam`)\n",
    "\n",
    "`samtools` to inspect, sort, merge, filter them\n",
    "\n",
    "`gatk MarkDuplicates` to mark or remove PCR duplicates from the data\n",
    "\n",
    "`mosdepth` to get coverage\n",
    "\n",
    "Genotype calling: what is the state of an individual at a given position in the genome?\n",
    "\n",
    "Program: `gatk HaplotypeCaller` - needs reference genome and mapped data\n",
    "\n",
    "Output: Variant Call File (`.vcf.gz`)\n",
    "\n",
    "`bcftools` as toolkit for inspection, filtering, etc.\n",
    "\n",
    "`bcftools view` - can include parameters for filtering, e.g. `-m2` means at least 2 alleles\n",
    "\n",
    "`bcftools concat` - concatenate files\n",
    "\n",
    "`bcftools stats` - obtain useful statistics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72718036",
   "metadata": {},
   "source": [
    "# VCF file formats\n",
    "\n",
    "We had generated two common types of VCF file outputs using the GATK `ERC` parameters: `BP_RESOLUTION` and `NONE`. One gives a genotype call for each position along the chromosome, the other only for positions that are different from the reference genome. The problem with the first is that it creates large files in which most positions are the same as the reference. The problem with the second is that we miss exactly where they are the same. A compromise is using the format `GVCF`, which creates a hybrid file format, in which consecutive `0/0` positions are merged.\n",
    "\n",
    "* If not done as part of the last challenge, you may create this file by using this parameter now, for the region 1-25000000.\n",
    "\n",
    "* Do you have it? Otherwise, we may have a look at it together.\n",
    "\n",
    "This format is very efficient for storing large data, but it creates a mess in terms of coordinates. At more advanced levels, other tools from the `gatk` (it's a toolkit) are necessary to handle this kind of information. Also, some commands of `bcftools` need to know what is going on, *e.g.* if you want to use this program to merge data from multiple individuals.\n",
    "\n",
    "* Now, also get the genotypes in the same format for any of the `.bam` files from the previous sessions!\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "cf1540e2",
   "metadata": {},
   "source": [
    "## Merging\n",
    "\n",
    "To clarify the difference between `bcftools concat` and `bcftools merge`: you concatenate different genomic positions for the same individuals, or your merge the same genomic positions for different individuals.\n",
    "\n",
    "**Concatenating:**\n",
    "![image-2.png](attachment:image-2.png)\n",
    "\n",
    "**Merging:**\n",
    "![image.png](attachment:image.png)\n",
    "\n",
    "\n",
    "Ok, then let's merge two individuals:\n",
    "\n",
    "```\n",
    "bcftools merge -Oz -o merged_set.vcf.gz -g /home/local/ANTHROPOLOGY/kuhlwilmm83/refgen/hg19/hg19.p13.plusMT.no_alt_analysis_set.fa.gz chr18_a.vcf.gz chr18_b.vcf.gz\n",
    "```\n",
    "\n",
    "* Let's have a look at the resulting file!\n",
    "\n",
    "Speaking about multiple individuals...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4bf7a77",
   "metadata": {},
   "source": [
    "## Diversity data\n",
    "\n",
    "### 1000 Genomes data\n",
    "\n",
    "The 1000 Genomes Project is a commonly-used reference data resources in human genetic studies.\n",
    "\n",
    "In the (final) phase 3 data, there are 2,504 individuals from 26 populations ([Sudmant et al. 2015](https://doi.org/10.1038/nature15394)).\n",
    "\n",
    "<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/e/ec/1000_Genomes_Project.svg/1280px-1000_Genomes_Project.svg.png?1657572184698\" alt=\"1KG\" width=\"1000\"/>\n",
    "\n",
    "**Figure 2 [Locations](https://www.wikiwand.com/en/1000_Genomes_Project#/Human_genome_samples) of population samples of 1000 Genomes Project.** Each circle represents the number of sequences in the final release.\n",
    "\n",
    "More information can be found in their [website](https://www.internationalgenome.org/).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c80cb124",
   "metadata": {},
   "source": [
    "# Population-scale VCFs\n",
    "\n",
    "## Extract data from a population\n",
    "\n",
    "So, we should get data from the 1000 Genomes project:\n",
    "\n",
    "```\n",
    "wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr18.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz\n",
    "```\n",
    "\n",
    "This is a multi-individual VCF file. This it not in the `gvcf` format, only SNPs are here, making it easier to handle.\n",
    "\n",
    "* Let's have a look at the header as well as the metadata file provided by the project!\n",
    "* Inspect the chromosome names!\n",
    "\n",
    "```\n",
    "bcftools view -h ALL.chr18.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz \n",
    "\n",
    "wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel\n",
    "less integrated_call_samples_v3.20130502.ALL.panel\n",
    "```\n",
    "\n",
    "\n",
    "### Reference problems\n",
    "\n",
    "Now, we encounter another typical problem - different [reference genomes](https://gatk.broadinstitute.org/hc/en-us/articles/360035891071-Reference-genome) for humans. A first problem is that existing data from different sources might have been mapped and genotype called using a newer version of the human genome ([version 38](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001405.26/): \"hg38\" from 2013), or the previous version ([version 19](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001405.13/) or \"hg19\" from 2009). These are the most common ones, and even older ones are really not recommended to use! These versions have different coordinates, so they are not compatible.\n",
    "\n",
    "To make things worse, although there is a coordinate system for the same version numbers, with the same number of bases from start to end, there exist versions with different chromosome names (\"chr18\" vs. (\"18\"), coming from different [sources](https://genome-euro.ucsc.edu/cgi-bin/hgGateway?db=hg19&redirect=manual&source=genome.ucsc.edu)!\n",
    "\n",
    "[Here](https://gatk.broadinstitute.org/hc/en-us/articles/360035890951-Human-genome-reference-builds-GRCh38-or-hg38-b37-hg19) is an explanation of the issue. That means, there are four different human genome versions commonly used (hg19 and hg38 with and without \"chr\").\n",
    "\n",
    "**When obtaining a dataset, you always need to check which reference genome in which format was used!**\n",
    "\n",
    "It's already a rather big file, can may see that it does contain >2M records for 2,504 individuals - a data frame of *>5.6 billion cells* for one of the smaller chromosomes... Even for the smallest human chromosome (chr21), this is >1M records!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d22a91f3",
   "metadata": {},
   "source": [
    "## Subsetting data\n",
    "\n",
    "### Subsetting regions\n",
    "\n",
    "* With this in mind, let's subset the file to make work easier! Note how the `-r` changes!\n",
    "\n",
    "```\n",
    "bcftools index ALL.chr18.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz \n",
    "bcftools view -r 18:1-25000000 -Oz -o chr18.vcf.gz ALL.chr18.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz\n",
    "bcftools view -h chr18.vcf.gz\n",
    "```\n",
    "\n",
    "* Let's calculate some stats:\n",
    "\n",
    "```\n",
    "bcftools stats chr18.vcf.gz > stats_chr18.vcf.gz\n",
    "bcftools index chr18.vcf.gz\n",
    "```\n",
    "\n",
    "The `index` step above is useful for retrieving subsets from the data, and generally recommendable for `vcf` files.\n",
    "\n",
    "### Subsetting individuals\n",
    "\n",
    "If we want to extract data from some individuals, for example, samples from the same population, we can use the argument `-S` with a file containing names of samples you want to extract. Let's assume you want to compare the Finns to the Iberians, you could use awk to create a nice list with the identifiers for each. (Btw, again a practical use case of a simple command line tool.)\n",
    "\n",
    "```\n",
    "awk '$2==\"FIN\" { print $1 }' integrated_call_samples_v3.20130502.ALL.panel > FIN.list\n",
    "awk '$2==\"IBS\" { print $1 }' integrated_call_samples_v3.20130502.ALL.panel > IBS.list\n",
    "```\n",
    "\n",
    "Now, each line in these files contains one sample ID. Then we can extract genetic variants from the FIN and IBS populations with the following commands, respectively.\n",
    "\n",
    "```\n",
    "bcftools view -Oz -o chr18.FIN.vcf.gz chr18.vcf.gz -S FIN.list &\n",
    "bcftools view -Oz -o chr18.IBS.vcf.gz chr18.vcf.gz -S IBS.list\n",
    "```\n",
    "\n",
    "* Now there is the `&` at the end of the line? What does this mean now?\n",
    "\n",
    "* Let's have a look at the files!\n",
    "\n",
    "As a result, you get a file containing the same variants as before, just for a subset of individuals instead of the whole thing.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae4c8c80",
   "metadata": {},
   "source": [
    "## Quality filtering\n",
    "\n",
    "### Subsetting SNPs by quality\n",
    "\n",
    "When analysing data, we usually want to use variants that can meet some conditions, for example, variants with good quality.\n",
    "\n",
    "* As before, we may extract biallelic SNPs that passed the quality checks.\n",
    "\n",
    "```\n",
    "bcftools view chr18.FIN.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.FIN.biallelic.snps.vcf.gz &\n",
    "bcftools view chr18.IBS.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.IBS.biallelic.snps.vcf.gz\n",
    "```\n",
    "\n",
    "If we have a look at these files, we can see that they contain a lot of positions where all individuals within the population are `0/0` (or `0|0`). If were are interested in what is going on within the population, we may restrict to polymorphisms within.\n",
    "\n",
    "\n",
    "### Restricting data to polymorphic sites\n",
    "\n",
    "Looking at the [documentation](https://samtools.github.io/bcftools/bcftools.html#view) of `bcftools view`, we find the flag `-a` to trim unseen alternative alleles. In this case, one should first trim the alleles, and afterwards filter for biallelic SNPs.\n",
    "\n",
    "* Let's trim and filter, and count how much is left in each case!\n",
    "\n",
    "```\n",
    "bcftools view chr18.FIN.vcf.gz -a | bcftools view -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.FIN.true_biallelic.snps.vcf.gz &\n",
    "bcftools view chr18.IBS.vcf.gz -a | bcftools view -f PASS -m 2 -M 2 -v snps | bgzip -c > chr18.IBS.true_biallelic.snps.vcf.gz\n",
    "```\n",
    "\n",
    "* Given that the number of individuals in each dataset is similar, does it tell us something about the diversity within these populations?\n",
    "\n",
    "\n",
    "### Much more filtering\n",
    "\n",
    "Beyond the very basic filtering above, `bcftools filter` provides a variety of options of filtering. The easiest are defined with `-e` to *exclude* sites with a certain condition, and `-i` to *include* those with a certain condition.\n",
    "\n",
    "```\n",
    "bcftools filter\n",
    "```\n",
    "\n",
    "You need to provide [*expressions*](https://samtools.github.io/bcftools/bcftools.html#expressions), in a very similar way you provide *expressions* to `awk`, including comparison operators (`>=` etc), logical operators (`&&`), selection of samples with square brackets (`[5:6]`), and many more which cannot be comprehensively discussed here.\n",
    "\n",
    "* Let's try some!\n",
    "\n",
    "```\n",
    "bcftools filter '-i GT[0]==\"0|1\"' chr18.vcf.gz | less\n",
    "\n",
    "bcftools filter '-e GT[20]==\"0|0\"' chr18.vcf.gz | less\n",
    "\n",
    "# to see what it is doing, look at:\n",
    "\n",
    "bcftools filter '-e GT[20]==\"0|0\"' chr18.vcf.gz | grep -v \"^#\" | cut -f 30 | less\n",
    "\n",
    "bcftools filter '-i GT[1]==\"1|2\"' chr18.vcf.gz | less\n",
    "```\n",
    "\n",
    "In the `INFO` column, there is a lot of information, such as allele frequencies (`AF`).\n",
    "\n",
    "```\n",
    "bcftools filter '-i INFO/AF>0.3' chr18.vcf.gz | less\n",
    "```\n",
    "\n",
    "\n",
    "The 1000 genomes file is particularly reduced, containing only genotypes. We may also try it on a more \"raw\" file from a previous session or the challenge.\n",
    "\n",
    "* Just to try out some:\n",
    "\n",
    "```\n",
    "bcftools filter '-i GT==\"./.\"' chr18_a.vcf.gz | less\n",
    "\n",
    "bcftools filter '-i FORMAT/DP>5' chr18_a.vcf.gz | less \n",
    "\n",
    "bcftools filter '-e FORMAT/DP<5 && FORMAT/DP>8' chr18_a.vcf.gz | less \n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6612541",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
