{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db7ed003-3088-48d4-a113-26e69b94a6d3",
   "metadata": {},
   "source": [
    "# Recap\n",
    "\n",
    "\n",
    "Raw sequencing data: `.fastq` format \n",
    "\n",
    "Removing adapters: `TRIMMOMATIC`\n",
    "\n",
    "Mapping: `bwa` sequencing data to `.bam`  format \n",
    "\n",
    "`samtools` program\n",
    "\n",
    "`samtools view` - look at (binary) file\n",
    "\n",
    "`samtools sort` - nicely sort reads along the chromosomes\n",
    "\n",
    "`samtools index` - create an index file\n",
    "\n",
    "`samtools stats` - create statistics for `.bam` file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf1540e2",
   "metadata": {},
   "source": [
    "# VCF file formats\n",
    "\n",
    "We have generated two common types of VCF file outputs using the GATK `ERC` parameters: `BP_RESOLUTION` and `NONE`. One gives a genotype call for each position along the chromosome, the other only for positions that are different from the reference genome. The problem with the first is that it creates large files telling us that most positions are the same as the reference. The problem with the second is that we miss exactly where they are the same. A compromise is the parameter `GVCF`, which creates a hybrid file format, in which the `0/0` positions are merged.\n",
    "\n",
    "If not done as part of the last challenge, you may create this file by using this parameter now!\n",
    "\n",
    "Otherwise, we may have a look at it together.\n",
    "\n",
    "This is very efficient for storing large data, but it creates a mess in terms of coordinates. Hence, other tools from the `gatk` (it's a toolkit by name) are necessary to handle this kind of information. Also, some commands of `bcftools` need to know what is going on, *e.g.* if you want to use this program to merge data from multiple individuals.\n",
    "\n",
    "Speaking about multiple individuals...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4bf7a77",
   "metadata": {},
   "source": [
    "## Diversity data\n",
    "\n",
    "### 1000 Genomes data\n",
    "\n",
    "The 1000 Genomes Project is a commonly-used reference data resources in human genetic studies.\n",
    "\n",
    "In the phase 3 data, there are 2,504 individuals from 26 populations ([Sudmant et al. 2015](https://doi.org/10.1038/nature15394)).\n",
    "\n",
    "<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/e/ec/1000_Genomes_Project.svg/1280px-1000_Genomes_Project.svg.png?1657572184698\" alt=\"1KG\" width=\"1000\"/>\n",
    "\n",
    "**Figure 2 Locations of population samples of 1000 Genomes Project.** Each circle represents the number of sequences in the final release. This figure is from https://www.wikiwand.com/en/1000_Genomes_Project#/Human_genome_samples.\n",
    "\n",
    "In this course, we usually use genotype data from the YRI and CEU populations as examples.\n",
    "\n",
    "Here,\n",
    "- YRI: Yoruba in Ibadan, Nigeria.\n",
    "- CEU: Utah Residents (CEPH) with Northern and Western European ancestry.\n",
    "\n",
    "More information can be found in their [website](https://www.internationalgenome.org/).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c80cb124",
   "metadata": {},
   "source": [
    "# More on bcftools\n",
    "\n",
    "`bcftools` is an effcient software for manipulating and analysing VCF files. There are several commands in `bcftools`.\n",
    "To find out available commands, you can type the following command in the terminal.\n",
    "\n",
    "```\n",
    "bcftools\n",
    "```\n",
    "\n",
    "In this tutorial, we will use `bcftools view` to look at genetic variants and extract some data from VCF files. More information about `bcftools` can be found in its [manual](https://samtools.github.io/bcftools/bcftools.html).\n",
    "\n",
    "## Check VCF files\n",
    "\n",
    "It is simple to view data stored in a VCF file. Just type `bcftools view <VCF file name>` in the terminal. \n",
    "\n",
    "Here, we provide an example VCF file named `chr21.YRI.CEU.vcf.gz`. You can type the following command in your terminal to have a look at our example data.\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.CEU.vcf.gz\n",
    "```\n",
    "\n",
    "As usual, it writes the whole file to the command line. Not good.\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.CEU.vcf.gz | less\n",
    "```\n",
    "\n",
    "With the argument `-H`, no header lines are printed out by `bcftools`. This is convenient when you have a VCF file with many header lines and just want to look at the variants.\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.CEU.vcf.gz -H | less\n",
    "```\n",
    "\n",
    "As you see, the file starts as a compressed file (.gz), but the output is uncompressed (you can read it).\n",
    "\n",
    "The common tool for this kind of file to compress it `bgzip`. Try to look at:\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.CEU.vcf.gz -H | bgzip -c | less\n",
    "```\n",
    "\n",
    "This you would usually write somewhere.\n",
    "\n",
    "How do you do that?\n",
    "\n",
    "\n",
    "## Extract data from a population\n",
    "\n",
    "If we want to extract data from some individuals, for example, samples from the same population, we can use the argument `-S` with a file containing names of samples you want to extract. \n",
    "\n",
    "Here, we provide two example files `YRI.list` and `CEU.list` with samples from the YRI and CEU populations.\n",
    "\n",
    "An example from `YRI.list` is below. Each line contains a sample name.\n",
    "\n",
    "```\n",
    "NA18486\n",
    "NA18488\n",
    "NA18489\n",
    "NA18498\n",
    "NA18499\n",
    "```\n",
    "\n",
    "Then we can extract genetic variants from the YRI and CEU populations with the following commands, respectively.\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.CEU.vcf.gz -S YRI.list | bgzip -c > chr21.YRI.vcf.gz\n",
    "bcftools view chr21.YRI.CEU.vcf.gz -S CEU.list | bgzip -c > chr21.CEU.vcf.gz\n",
    "```\n",
    "\n",
    "Let's have a look at the files!\n",
    "\n",
    "\n",
    "## Filtering data\n",
    "\n",
    "When analysing data, we usually want to use variants that can meet some conditions, for example, variants with good quality. Then we can use several arguments to filter data in VCF files.\n",
    "\n",
    "For example, the following commands extract biallelic single nucleotide polymorphisms (SNPs) that passed quality checks from the YRI and CEU populations.\n",
    "\n",
    "```\n",
    "bcftools view chr21.YRI.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > chr21.YRI.biallelic.snps.vcf.gz\n",
    "bcftools view chr21.CEU.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > chr21.CEU.biallelic.snps.vcf.gz\n",
    "```\n",
    "\n",
    "What is the meaning of the arguments `-f`, `-m`, `-M`, and `-v`? Ask bcftools directly: `bcftools view -h` \n",
    "\n",
    "Or use the [manual](https://samtools.github.io/bcftools/bcftools.html#view)!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
