{"metadata":{"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kernelspec":{"name":"python3","display_name":"Python 3 (ipykernel)","language":"python"}},"nbformat_minor":5,"nbformat":4,"cells":[{"cell_type":"markdown","source":"## Introduction\n\n### VCF files\n\nThe variant call format (VCF) is a standard format of a text file to store genetic variants and their metadata. An example of a VCF file is shown below.\n\n<img src=\"https://davetang.github.io/learning_vcf_file/img/vcf_format.png\" alt=\"VCF example\" width=\"1000\"/>\n\n**Figure 1 A VCF file.** Lines started with the character `#` are header lines. These header lines define different metadata. Lines started without `#` store genetic variants. Each line represents a variant. This figure is from https://davetang.github.io/learning_vcf_file/#introduction.\n\n\nMore information about VCF files can be found in [hts-specs](https://samtools.github.io/hts-specs/).\n\n### The 1000 Genomes Project\n\nThe 1000 Genomes Project is a commonly-used reference data resources in human genetic studies.\n\nIn the phase 3 data, there are 2,504 individuals from 26 populations ([Sudmant et al. 2015](https://doi.org/10.1038/nature15394)).\n\n<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/e/ec/1000_Genomes_Project.svg/1280px-1000_Genomes_Project.svg.png?1657572184698\" alt=\"1KG\" width=\"1000\"/>\n\n**Figure 2 Locations of population samples of 1000 Genomes Project.** Each circle represents the number of sequences in the final release. This figure is from https://www.wikiwand.com/en/1000_Genomes_Project#/Human_genome_samples.\n\nIn this course, we usually use genotype data from the YRI and CEU populations as examples.\n\nHere,\n- YRI: Yoruba in Ibadan, Nigeria.\n- CEU: Utah Residents (CEPH) with Northern and Western European ancestry.\n\nMore information can be found in their [website](https://www.internationalgenome.org/).\n\n## bcftools usage\n\n`bcftools` is an effcient software for manipulating and analysing VCF files. There are several commands in `bcftools`.\nTo find out available commands, you can type the following command in the terminal.\n\n```\nbcftools\n```\n\nIn this tutorial, we will use `bcftools view` to look at genetic variants and extract some data from VCF files. More information about `bcftools` can be found in its [manual](https://samtools.github.io/bcftools/bcftools.html).\n\n### Check VCF files\n\nIt is simple to view data stored in a VCF file. Just type `bcftools view <VCF file name>` in the terminal. \n\nHere, we provide an example VCF file named `./data/chr21.YRI.CEU.vcf.gz`. You can type the following command in your terminal to have a look at our example data.\n\n```\nbcftools view ./data/chr21.YRI.CEU.vcf.gz\n```\n\nBy default, `bcftools` outputs all the data from the file into the terminal.\n\nHowever, this is slow and not easy to look at the data. Therefore, we can modify the command as follow.\n\n```\nbcftools view ./data/chr21.YRI.CEU.vcf.gz | less -S\n```\n\nThe character `|` transfers the output from `bcftools` as the input for another tool called `less`.\n\n`less` is a Linux command that can read a text file and display its content one page at a time. More information can be found in its [manual](https://man7.org/linux/man-pages/man1/less.1.html).\n\nWith the argument `-H`, no header lines are printed out by `bcftools`. This is convenient when you have a VCF file with many header lines and just want to look at the variants.\n\n```\nbcftools view ./data/chr21.YRI.CEU.vcf.gz -H | less -S\n```\n\n### Extract data from a population\n\nIf we want to extract data from some individuals, for example, samples from the same population, we can use the argument `-S` with a file containing names of samples you want to extract. \n\nHere, we provide two example files `./data/YRI.list` and `./data/CEU.list` with samples from the YRI and CEU popualtions.\n\nAn example from `./data/YRI/list` is below. Each line contains a sample name.\n\n```\nNA18486\nNA18488\nNA18489\nNA18498\nNA18499\n```\n\nThen we can extract genetic variants from the YRI and CEU populations with the following commands, respectively.\n\n```\nbcftools view ./data/chr21.YRI.CEU.vcf.gz -S ./data/YRI.list | bgzip -c > ./data/chr21.YRI.vcf.gz\nbcftools view ./data/chr21.YRI.CEU.vcf.gz -S ./data/CEU.list | bgzip -c > ./data/chr21.CEU.vcf.gz\n```\n\nHere, `bgzip -c` can compress the extracted data and help us to save some space. \n\nAnd the character `>` redirects the output from the terminal into a file, such as `./data/chr21.YRI.vcf.gz` and `./data/chr21.CEU.vcf.gz`.\n\n### Filter data\n\nWhen analysing data, we usually want to use variants that can meet some conditions, for example, variants with good quality. Then we can use several arguments to filter data in VCF files.\n\nFor example, the following commands extract biallelic single nucleotide polymorphisms (SNPs) that passed quality checks from the YRI and CEU populations.\n```\nbcftools view ./data/chr21.YRI.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > ./data/chr21.YRI.biallelic.snps.vcf.gz\nbcftools view ./data/chr21.CEU.vcf.gz -f PASS -m 2 -M 2 -v snps | bgzip -c > ./data/chr21.CEU.biallelic.snps.vcf.gz\n```\nYou can find out the meanings of the arguments `-f`, `-m`, `-M`, and `-v` by typing `bcftools view -h` in the terminal or from the [manual](https://samtools.github.io/bcftools/bcftools.html#view).","metadata":{"tags":[]},"id":"7359d202-bf95-4dea-be9b-75d0bbb15a32"}]}